packages/protocol/src/utils/event.ts

```diff
@@ -0,0 +1,92 @@
+/********************************************************************************
+ * Copyright (c) 2023 EclipseSource and others.
+ *
+ * This program and the accompanying materials are made available under the
+ * terms of the Eclipse Public License v. 2.0 which is available at
+ * http://www.eclipse.org/legal/epl-2.0.
+ *
+ * This Source Code may also be made available under the following Secondary
+ * Licenses when the conditions for such availability set forth in the Eclipse
+ * Public License v. 2.0 are satisfied: GNU General Public License, version 2
+ * with the GNU Classpath Exception which is available at
+ * https://www.gnu.org/software/classpath/license.html.
+ *
+ * SPDX-License-Identifier: EPL-2.0 OR GPL-2.0 WITH Classpath-exception-2.0
+ ********************************************************************************/
+import * as jsonrpc from 'vscode-jsonrpc';
+import { Disposable } from './disposable';
+import { SafeFunction } from './type-util';
+
+/**
+ * GLSP event/emitter API. Reuses the API provided by
+ * `vscode-jsonrpc` and overrides certain type definitions to match the type
+ * definitions provided by GLSP (Event, Disposable)
+ */
+
+/**
+ * Represents a typed event. Listener functions can subscribe for a event
+ * and will be called when the event is fired. Typically events are provided by services via
+ * public property. For example assume we have a service `myService` that exposes a
+ * `onStartupComplete` event. Then a listener could be registered like this
+ * ```ts
+ *      const disposable= this.myService.onStartupComplete(event=> doSomething(event));
+ *
+ *     // If the listener should be unregistered again, simply call
+ *      disposable.dispose()
+ * ```
+ */
+export interface Event<T> extends jsonrpc.Event<T> {
+    /**
+     *
+     * @param listener The listener function will be called when the event happens.
+     * @param thisArgs The 'this' which will be used when calling the event listener.
+     * @param disposables An array to which the {@link Disposable} for removing the listener will be added.
+     * @returns a {@link Disposable} to remove the listener again.
+     */
+    (listener: (e: T) => any, thisArgs?: any, disposables?: Disposable[]): Disposable;
+}
+/**
+ * Optional options that can be passed to the constructor
+ * of an {@link Emitter}.
+ */
+export interface EmitterOptions {
```

Comment:
Should that extends the `EmitterOptions` from jsonrpc? Cause otherwise the constructor might break at some point.

Reply:
Unfortunatley the `EmitterOptions` are not part of the main export of `vscode-jsonrpc`.

Reply:
Got it!